### PR TITLE
Add `input_device` to `StandardUpdater`

### PR DESCRIPTION
**Pull request by niboshi**
Replaces #7324

Required in #7330

### CODE REVIEW

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py [hunk 1]
```diff
@@ -115,6 +118,257 @@ def test_serialize(self):
...
+        if expected_model_device is None:
+            # Model device is unchanged
+            # TODO(niboshi): model.device should be initial_model_device too.
+            assert model.p1.device == initial_model_device
+            assert model.p2.device == initial_model_device
+        else:
+            # Model is transferred
+            # TODO(niboshi): model.device should be expected_model_device too.
+            assert model.p1.device == expected_model_device
+            assert model.p2.device == expected_model_device
```
Comment (hvy) [inline tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: I minor but you can DRY this by conditionally assigning the RHS to a temporary.

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py [hunk 2]
```diff
@@ -115,6 +118,257 @@ def test_serialize(self):
...
+            input_device_arg):
+        # Determines the expected devices.
+        # Returns: (
+        #    expected_device_attr:   Expected StandardUpdater.device
+        #    expected_model_device:  Expected model device
+        #    expected_input_device:  Expected device given to converters
+        # )
+        # or None if an error is expected.
+
+        if device_arg is not None:
```
Comment (hvy) [inline tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: Do you need this condition (you can reduce one indentation level and also think it's easier to read)?

File: chainer/training/updaters/standard_updater.py [hunk 3]
```diff
@@ -78,19 +99,22 @@ def __init__(self, iterator, optimizer, converter=convert.concat_examples,
...
         self._optimizers = optimizer
 
+        # Transfer the model
         if device is not None:
             for optimizer in six.itervalues(self._optimizers):
                 if device.xp is cuda.cupy:
                     # Do not transfer between different cupy devices.
-                    # TODO(niboshi): Reconsider this behavior
+                    # TODO(niboshi): Detect GPU-to-GPU transfer and raise
+                    # FutureWarning. Eventually replace it with to_device.
```
Comment (hvy) [inline chainer/training/updaters/standard_updater.py]: Is this TODO to be addressed in a separate PR? Should we create an issue if so be the case?

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py [hunk 4]
```diff
@@ -115,6 +118,257 @@ def test_serialize(self):
...
+
+            # If device_arg is GpuDevice,  it will skip GPU-to-GPU transfer of
+            # the model (but not input).
+            if isinstance(device_arg, backend.GpuDevice):
+                if isinstance(initial_model_device, backend.GpuDevice):
+                    expected_model_device = initial_model_device
+                else:
+                    expected_model_device = device_arg
+
+                expected_input_device = device_arg
```
Comment (hvy) [inline tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: Should the RHS not be `input_device_arg`?

File: chainer/training/updaters/standard_updater.py [hunk 5]
```diff
@@ -48,12 +53,19 @@ class StandardUpdater(_updater.Updater):
             :meth:`~chainer.Optimizer.new_epoch` of the main optimizer is
             automatically called when the ``is_new_epoch`` attribute of the
             main iterator is ``True``.
+        input_device (device specifier):
+            Device to which the training data is sent.
+            If ``input_device`` is omitted and ``model_device`` is given,
```
[the remark this answers is not in the document]
Comment (niboshi) [reply chainer/training/updaters/standard_updater.py]: Fixed.

File: chainer/training/updaters/standard_updater.py [hunk 6]
```diff
@@ -48,12 +53,19 @@ class StandardUpdater(_updater.Updater):
...
+            ``input_device`` will match the ``device`` argument.
+            If both omitted, the device of the input will stay unchanged.
+            Also see the note below.
```
[the remark this answers is not in the document]
Comment (niboshi) [reply chainer/training/updaters/standard_updater.py]: Removed this sentence.

File: chainer/training/updaters/standard_updater.py [hunk 7]
```diff
@@ -102,6 +126,14 @@ def __init__(self, iterator, optimizer, converter=convert.concat_examples,
             for o in six.itervalues(self._optimizers):
                 o.use_auto_new_epoch = True
 
+    @property
+    def device(self):
+        return self._device
```
[the remark this answers is not in the document]
Comment (niboshi) [reply chainer/training/updaters/standard_updater.py]: Yes, it's intentional.
Modifying the attribute will not work as might be expected.
As the attribute itself is documented, we can't remove it (and I don't think there will never be any useful cases).

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py [hunk 8]
```diff
@@ -115,6 +118,257 @@ def test_serialize(self):
...
+
+
+@chainer.testing.backend.inject_backend_tests(None, _backend_params)
+@chainer.testing.backend.inject_backend_tests(None, _backend_params)
+@chainer.testing.backend.inject_backend_tests(None, _backend_params)
+class TestStandardUpdaterDevice(unittest.TestCase):
+
+    def test_device(
+            self, model_initial_backend_config, model_backend_config,
+            input_backend_config):
```
[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: `model_backend_config` could be `device_backend_config`, as (partly) suggested.
I think "initial" has some important information. Don't you think so?

`backend_config` could be replaced with `backend` as suggested, but that convention is never used anywhere else (to my knowledge).

File: chainer/training/updaters/standard_updater.py (hunk 3 above)
Comment (niboshi) [reply chainer/training/updaters/standard_updater.py]: Yes, it requires quite an implementation. It won't be done in this PR.

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py (hunk 4 above)
Comment (niboshi) [reply tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: Thanks. Fixed and added more tests.

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py (hunk 2 above)
Comment (niboshi) [reply tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: Fixed.

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py (hunk 1 above)
Comment (niboshi) [reply tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: Fixed.

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py [hunk 9]
```diff
@@ -115,6 +118,257 @@ def test_serialize(self):
...
+        #    expected_device_attr:   Expected StandardUpdater.device
+        #    expected_model_device:  Expected model device
+        #    expected_input_device:  Expected device given to converters
+        # )
+        # or None if an error is expected.
+
+        if device_arg is not None:
+            # device argument is given
+
+            # If device_arg is GpuDevice,  it will skip GPU-to-GPU transfer of
```
[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: Fixed.

**Comment by hvy** [comment]
By the way, should `ParallelUpdater` follow this style too (not sure about `MultiprocessParallelUpdater`)?

File: chainer/training/updaters/standard_updater.py (hunk 3 above)
Comment (hvy) [reply chainer/training/updaters/standard_updater.py]: Created a tracking issue since this is rather a long-termed process https://github.com/chainer/chainer/issues/7504.

**Comment by hvy** [comment]
Jenkins, test this please.

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: Sorry, those suggestions above was meant to all be suffixed with `_config`. But it's not a strong opinion and I think it's fine as is.

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py [hunk 10]
```diff
@@ -115,6 +118,283 @@ def test_serialize(self):
...
+        # Check the model device.
+        assert model.device == device
+
+        updater.update_core()
+
+        assert optimizer.update.call_count == 1
+        args, kwargs = optimizer.update.call_args
+        assert len(args) == 2
+        assert len(kwargs) == 0
+        loss, v1 = args
```
Comment (hvy) [inline tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: Some minors but how about using a `_` for `loss` as it's intensionally not used and different name for `v1`, .e.g. `in_array`.

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py [hunk 11]
```diff
@@ -115,6 +118,283 @@ def test_serialize(self):
...
+            device_arg=DummyDevice(0),
+            input_device_arg=backend.GpuDevice.from_device_id(1))
+
+        # device      : omitted
+        # input_device: GPU 1
+        self.check_device_argument_fallback(
+            initial_model_device=initial_model_device,
+            initial_input_device=initial_input_device,
+            device_arg=None,
+            input_device_arg=backend.GpuDevice.from_device_id(1))
```
Comment (hvy) [inline tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: As these checks may fail independently, how about separating them?

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py (hunk 10 above)
Comment (niboshi) [reply tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: They're consistent with other tests.

File: tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py (hunk 11 above)
Comment (niboshi) [reply tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py]: These tests are closely related.
Separating them would make that point less apparent.
I wanted to write this test with parameterization but it's not possible so I wrote in this manner.

**Comment by hvy** [comment]
LGTM!
